### ISSUE TITLE
Couple breaking issues using opencv4nodejs with electron.

### ISSUE THREAD

**Issue by wbern** [justadudewhohacks/opencv-electron#13]
Hi. I'm using electron-vue, and despite a lot of troubleshooting (even attaching to the electron main process), I am getting devtool disconnects that are both random and consequent.

Here is a list of functions that I've found to cause breakage. I haven't gotten any further programming, but I hope this can be enough to make opencv4nodejs more stable as I love this wrapper!

```
mounted() {
    // issues are the same no matter if you use camera or video
    // this.wCap = new cv.VideoCapture(1)
    this.wCap = new cv.VideoCapture(path.resolve('../backend/media/video.mp4'))

    this.wCap.readAsync(frame => {
        console.log(frame) // null
    })

    // console.log(this.wCap.read()) // crashes devtools in electron, is VideoCapture not ready?

    // setInterval(() => {
    //     console.log(this.wCap.read()) // works ok, returns:
    //     // "{"step":3840,"elemSize":3,"sizes":[720,1280],"empty":false,"depth":0,"dims":2,"channels":3,"type":16,"cols":1280,"rows":720}"
    // }, 500)

    // setInterval(() => {
    //     console.log(this.wCap.read()) // too fast? returns a lot of the following
    //     // "{"step":0,"elemSize":0,"sizes":[],"empty":true,"depth":0,"dims":0,"channels":1,"type":0,"cols":0,"rows":0}"
    // }, 10)

    setInterval(() => {
        let frame = this.wCap.read()
        // below causes crashes randomly and disconnects devtools, no matter the interval (tested 500 and 2000)
        // zero error output, not sure where to see the errors
        let base64 = cv.imencode('.png', frame).toString('base64')
    }, 500)
},
```

**Versions**
VS Code version: 1.21
Electron version: 1.8.7
Chromium: v59
Node: 8.2.1
opencv4nodejs: 3.5.0
opencv: prebuilt windows, 3.4.1

I have tried rebuilding opencv4nodejs to electron.

**Comment by wbern** [comment]
I just tested opencv4nodejs 4.5.1 as well, with the same results.

**Comment by wbern** [comment]
Ok, I have just learned that the reason devtools disconnects is because I'm running nodejs code inside the Vue instances. That is interesting to know.

I've also learned that `cv.imencode('.png', frame).toString('base64')` is crashing because it's receiving one of those strange 1-channel frames that I documented as a comment in the code, in the first post.

Not sure why readAsync returns null, and why read() returns some kind of empty frame occassionally.

This is going to make electron usage tough until this is resolved in my case. :(

**Comment by justadudewhohacks** [comment]
Devtools probably disconnect because your nodejs process crashes?

If you are using async methods with callbacks, keep in mind that the first cb arg is an error or null and the second one the result:
``` javascript
this.wCap.readAsync((error, frame) => {
...
})
```

It is possible that `read()` returns an empty Mat if it could not read a frame. You should check for that in your code and if `mat.empty` is true, skip the frame.

**Comment by wbern** [comment]
> If you are using async methods with callbacks, keep in mind that the first cb arg is an error or null and the second one the result:

Wow, I missed the fact that there's an error parameter... Thanks!

> Devtools probably disconnect because your nodejs process crashes?

I suspected the nodejs process too. I'm using this repo. https://github.com/SimulatedGREG/electron-vue

When I start it up, 5 electron.exe processes are spawned (probably fine according to [source](https://github.com/electron/electron/issues/2437#issuecomment-128416641)), and when I trigger an opencv crash, one of the Electron.exe's go away with zero error messages.

If I put the same code in the renderer process but _not_ within the Vue instances, I get error messages to console, and chrome devtool does not disconnect either.

**Comment by justadudewhohacks** [comment]
I think remote processes, like node processes, don't handle uncaught excepetions, they simply shut down. The renderer process simply logs uncaught exceptions to the console, like in the browser.

But you should be able to catch all excepetions from opencv4nodejs, unless they are segfaults.

**Comment by wbern** [comment]
I've tried debugging both the main process, and the dev-runner script, which essentially runs the renderer through the webpack dev server.

I'm starting to wonder if I need to find a way to troubleshoot what happens inside this:
```
let app = require('electron').app
```

**Comment by wbern** [comment]
I'm close but no success so far.

```

    mainWindow.webContents.on('crashed', (event, sender) => {
        console.log(event) // just returns a function called preventDefault()
    })
```

This gives me a callback when the site crashes, but no stack trace.

Final update: When opencv crashes, it seems to crash the entire browser. I can get small errors to show up properly in the console often now, but every now and then, readAsync and read() together with imencode, will simply crash the browser.

```
mounted() {
                this.wCap.readAsync((err, frame) => {
                    if (frame && !frame.empty) {
                        // crashes the browser intermittently.
                        let base64 = cv.imencode('.png', frame).toString('base64')
                    }
                })
}
```

I'm at a loss for how to proceed after spending 6+ hours on debugging.

**Comment by justadudewhohacks** [comment]
Oh sorry, now that you mention readAsync again, I just commented in #12 why this causes crashes in electron. Why imencode crashes I am not quite sure though.

**Comment by wbern** [comment]
@justadudewhohacks it's really amazing that you're supporting electron so well for a library like this one. I will await the fix and start building my application in the mean time.

PS. I am scheduled to write a blog post of my endeavors with this library within a month or so, I hope that counts for something.

**Comment by justadudewhohacks** [comment]
Nice I would love to read and share your blog post, once it's released :). Let me know

**Comment by justadudewhohacks** [comment]
Apparently the node buffer issue I once had with `mat.getData()` is not apparent with `VideoCaptrure.read`. I can read frames from a video file without any issues in the renderer thread. I think your problem could be, you are using setInterval, which might lead to race conditions with `readAsync` between different threads.

Atleast I found my app crashing using `setInterval` + `readAsync`. Could you try it again by using `setTimeout` after reading a frame instead of `setInterval`? Consider the following code snippet:
``` javascript
async function readFrame() {
  let frame = await cap.readAsync()
  if (!frame.empty) {
    cv.imshow('cap', frame)
    cv.waitKey(1)
  }
  setTimeout(readFrame, 0)
}
readFrame()
```

**Comment by wbern** [comment]
@justadudewhohacks I know that I did not use setInterval when I put the same code in a vue onClick callback for a vue button, and I was having intermittent issues then still, no matter how long I waited between clicks. It looked almost completely random. Perhaps vue is doing something asynchronously that's no good then?

**Comment by justadudewhohacks** [comment]
Hmm, with an onClick handler you could still potentially run in race conditions with using `readAsync`, without checking whether a thread is still reading a frame. I don't see a reason why vue is doing something async, that would cause this issue.

What's odd though, I would have expected to encounter the node buffer issue here again, but currently can't reproduce it with the following setup:
```
node: 9.2.0
electron: 1.8.4
electron-rebuild: 1.7.3
opencv4nodejs: 4.5.1
```

I just hope that your (5?) vue processes are not trying to access the same resource/ file at the same time? Could you maybe still try the above shown sample code in your environment and let it run for some time to check what the actual issue is?

**Comment by wbern** [comment]
@justadudewhohacks in my previous post, I had the same issues with read(), if that makes a difference.

**Comment by justadudewhohacks** [comment]
Okay I think I kind of confused your issue with #12 . You already said that the reason for the application crash actually is `imencode` receiving single channeled or empty Mats, it's not that `read` in itself crashes your app right? Correct me if I am wrong.

Besides checking the frame for being empty, as you already do, I would also check the number of channels and in case it is a gray image convert it to a 3 channel image: `const frame3C = frame.channels === 1 ? frame.cvtColor(cv.COLOR_GRAY2BGR) : frame`.

**Comment by wbern** [comment]
I have added you as part of my private repo.

I have opencv installed globally. You are correct, imencode is causing the crashes.

The fix you suggested does not help (with frame channels). It appears there's always 3 channels in my current approach.

The most annoying thing of all is that I'm not getting any error information whatsoever to share with you from electron. devtool just disconnects..

**Comment by wbern** [comment]
New find: All the devtool crashing happens inside a BrowserWindow instance.

**Comment by justadudewhohacks** [comment]
I figured out the issue. I confused quite a bit, actually `cv.imencode` returns a node buffer and not `cap.read()`, which causes the exact same node buffer issue as I initially said. 

I can reproduce it now, the crash occurs exactly when the buffer returned from `cv.imencode` gets garbage collected. Fix coming soon.

**Comment by wbern** [comment]
Awesome!!

**Comment by justadudewhohacks** [comment]
Fix is available in v4.5.2 now.

**Comment by wbern** [comment]
> Fix is available in v4.5.2 now.

Great news, your fix works for me. Using cv.read(), I spammed a button in vue to capture 2 different cameras synchronously. the UI thread hung due to the sheer amount of requests, but when I was done, it was working just fine. I also tested using setInterval using both read() and readAsync() and it worked fine as well.

Closing!
